In [ ]:
Contents
1. Import libraries
2. Import data
3. Create a subset
4. Define a function
5. Apply the function
6. Creating new variable 'busiest days'
7. Name a column for hours of the day

01. Importing libraries

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

02. Importing data

In [2]:
#Importing data
df_orders_products_merged = pd.read_pickle(r"C:\Users\kelly\OneDrive\Desktop\Prepared Data\orders_products_merged.pickle")

In [6]:
df_orders_products_merged.shape

(30328763, 18)

In [9]:
#Create a subset
df=df_orders_products_merged[:1000000]

In [10]:
df.shape

(1000000, 18)

In [14]:
#Define a function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [16]:
#Apply the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\kelly\AppData\Local\Temp\ipykernel_19200\3959542531.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [17]:
df['price_range'].value_counts(dropna = False)

Mid-range product    631969
Low-range product    368031
Name: price_range, dtype: int64

In [18]:
df['prices'].max()

14.0

In [19]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\kelly\AppData\Local\Temp\ipykernel_19200\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [20]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [21]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [22]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    631969
Low-range product    368031
Name: price_range_loc, dtype: int64

In [24]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [25]:
df_orders_products_merged.loc[(df_orders_products_merged['prices'] <= 15) & (df_orders_products_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [26]:
df_orders_products_merged.loc[df_orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [27]:
df_orders_products_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     20462144
Low-range product      9476774
High-range product      389845
Name: price_range_loc, dtype: int64

In [30]:
df_orders_products_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30328763 entries, 0 to 30328762
Data columns (total 19 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   Unnamed: 0.1            int64   
 1   Unnamed: 0_x            int64   
 2   order_id                int64   
 3   user_id                 int64   
 4   eval_set                object  
 5   order_number            int64   
 6   order_dow               int64   
 7   order_hour_of_day       int64   
 8   days_since_prior_order  float64 
 9   product_id              int64   
 10  add_to_cart_order       int64   
 11  reordered               int64   
 12  _merge                  category
 13  Unnamed: 0_y            int64   
 14  product_name            object  
 15  aisle_id                int64   
 16  department_id           int64   
 17  prices                  float64 
 18  price_range_loc         object  
dtypes: category(1), float64(2), int64(13), object(3)
memory usage: 4.3+ GB


In [31]:
df_orders_products_merged.rename(columns = {'order_dow': 'orders_day_of_week'}, inplace = True)

In [32]:
df_orders_products_merged.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0             1             1   2398795  2398795    prior             2   
1             2             2    473747   473747    prior             3   
2             3             3   2254736  2254736    prior             4   
3             4             4    431534   431534    prior             5   
4             5             5   3367565  3367565    prior             6   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   3                  7                    15.0         196   
1                   3                 12                    21.0         196   
2                   4                  7                    29.0         196   
3                   4                 15                    28.0         196   
4                   2                  7                    19.0         196   

   add_to_cart_order  reordered _merge  Unnamed: 0_y product_name  aisle_id  \
0                  1          1   both           195         Soda        77   
1                  1          1   both           195         Soda        77   
2                  1          1   both           195         Soda        77   
3                  1          1   both           195         Soda        77   
4                  1          1   both           195         Soda        77   

   department_id  prices    price_range_loc  
0              7     9.0  Mid-range product  
1              7     9.0  Mid-range product  
2              7     9.0  Mid-range product  
3              7     9.0  Mid-range product  
4              7     9.0  Mid-range product

In [33]:
df_orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    5779087
1    5303718
6    4190948
5    3952326
2    3947564
3    3600589
4    3554531
Name: orders_day_of_week, dtype: int64

In [34]:
result = []

for value in df_orders_products_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [35]:
result

['Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Reg

In [36]:
df_orders_products_merged['busiest_day'] = result

In [37]:
df_orders_products_merged['busiest_day'].value_counts (dropna = False)

Regularly busy    20995145
Busiest day        5779087
Least busy         3554531
Name: busiest_day, dtype: int64

In [39]:
df_orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    5779087
1    5303718
6    4190948
5    3952326
2    3947564
3    3600589
4    3554531
Name: orders_day_of_week, dtype: int64

In [58]:
#2 Creating new variable 'busiest days'
result2 = []

for value in df_orders_products_merged["orders_day_of_week"]:
  if value in [0,1]:
    result2.append("Busiest days")
  elif value in [3,4]:
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")

In [70]:
result2

['Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busies

In [59]:
df_orders_products_merged['busiest days'] = result2

In [60]:
df_orders_products_merged['busiest days'].value_counts(dropna = False)

Regularly busy    12090838
Busiest days      11082805
Slowest days       7155120
Name: busiest days, dtype: int64

#3 Check the values of this new column for accuracy.  
The total number of busiest days is the sum of 0 & 1 (11082805) and slowest days is the sum of 3 & 4 (7155120).

#4 Determine the busiest hours of the day

In [65]:
#Name of column for hours of the day
df_orders_products_merged.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0             1             1   2398795  2398795    prior             2   
1             2             2    473747   473747    prior             3   
2             3             3   2254736  2254736    prior             4   
3             4             4    431534   431534    prior             5   
4             5             5   3367565  3367565    prior             6   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   3                  7                    15.0         196   
1                   3                 12                    21.0         196   
2                   4                  7                    29.0         196   
3                   4                 15                    28.0         196   
4                   2                  7                    19.0         196   

   ...  _merge  Unnamed: 0_y product_name  aisle_id department_id  prices  \
0  ...    both           195         Soda        77             7     9.0   
1  ...    both           195         Soda        77             7     9.0   
2  ...    both           195         Soda        77             7     9.0   
3  ...    both           195         Soda        77             7     9.0   
4  ...    both           195         Soda        77             7     9.0   

     price_range_loc     busiest_day    busiest days    busiest_days  
0  Mid-range product  Regularly busy    Slowest days  Regularly days  
1  Mid-range product  Regularly busy    Slowest days  Regularly days  
2  Mid-range product      Least busy    Slowest days  Regularly days  
3  Mid-range product      Least busy    Slowest days  Regularly days  
4  Mid-range product  Regularly busy  Regularly busy  Regularly days  

[5 rows x 22 columns]

In [66]:
#Hour of day values
df_orders_products_merged['order_hour_of_day'].value_counts()

10    2593725
11    2564597
14    2517238
15    2487586
13    2487500
12    2445841
16    2364969
9     2311334
17    1943858
8     1622394
18    1520954
19    1169224
20     910005
7      844665
21     746254
22     592432
23     375889
6      274801
0      203460
1      108110
5       82706
2       63961
4       49400
3       47860
Name: order_hour_of_day, dtype: int64

In [69]:
result3 = []

for value in df_orders_products_merged['order_hour_of_day']:
  if value >= 9 and value <=16:
    result3.append("Most Orders")
  elif value <= 6:
    result3.append("Fewest Orders")
  else:
    result3.append("Average Orders")

In [71]:
result3

['Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders

In [73]:
df_orders_products_merged["busiest_period_of_day"] = result3

In [74]:
df_orders_products_merged["busiest_period_of_day"].value_counts(dropna = False)

Most Orders       19772790
Average Orders     9725675
Fewest Orders       830298
Name: busiest_period_of_day, dtype: int64

In [78]:
df_orders_products_merged.to_pickle(r'C:\Users\kelly\OneDrive\Desktop\Prepared Data/orders_products_merged_v2.pkl')